drivers/tee/optee/optee.c

```diff
@@ -1129,23 +1130,89 @@ static int set_optee_method(void)
 	return 0;
 }
 
+static bool optee_check_uid(const struct device *dev)
+{
+	struct arm_smccc_res res;
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+
+	data->smc_call(OPTEE_SMC_CALLS_UID, 0, 0, 0, 0, 0, 0, 0, &res);
+
+	if (res.a0 == OPTEE_MSG_UID_0 && res.a1 == OPTEE_MSG_UID_1 &&
+	    res.a2 == OPTEE_MSG_UID_2 && res.a3 == OPTEE_MSG_UID_3) {
+		return true;
+	}
+
+	return false;
+}
+
+static void optee_get_revision(const struct device *dev)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	union {
+		struct arm_smccc_res smccc;
+		struct optee_smc_call_get_os_revision_result result;
+	} res = { 0 };
+
+	data->smc_call(OPTEE_SMC_CALL_GET_OS_REVISION, 0, 0, 0, 0, 0, 0, 0,
+		  &res.smccc);
+
+	if (res.result.build_id) {
+		LOG_INF("OPTEE revision %lu.%lu (%08lx)", res.result.major,
+			res.result.minor, res.result.build_id);
+	} else {
+		LOG_INF("OPTEE revision %lu.%lu", res.result.major, res.result.minor);
+	}
+}
+
+static bool optee_exchange_caps(const struct device *dev, uint32_t *sec_caps)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	union {
+		struct arm_smccc_res smccc;
+		struct optee_smc_exchange_capabilities_result result;
+	} res;
+	uint32_t a1 = 0;
+
+	if (!IS_ENABLED(CONFIG_SMP) || arch_num_cpus() == 1) {
+		a1 |= OPTEE_SMC_NSEC_CAP_UNIPROCESSOR;
+	}
+
+	data->smc_call(OPTEE_SMC_EXCHANGE_CAPABILITIES, a1, 0, 0, 0, 0, 0, 0,
+		  &res.smccc);
+
+	if (res.result.status != OPTEE_SMC_RETURN_OK) {
+		return false;
+	}
+
+	*sec_caps = res.result.capabilities;
```

Comment:
This is out of scope of the task I was given.

Reply:
I though you should make a second part of this issue https://github.com/xen-troops/zephyr/issues/61
so, will there be a second part of changes related to capabilities ?

Reply:
There should be multiple level capabilities handling in the optee feature. First is driver level and the second is in client level with extra get_version syscall. Those features are not implemented for now. But I think Zephyr should know which capabilities it supports. For example there is no implementation of shared memory pool usage. I think it will not take long to take a look through all capabilities and make some checks.

Reply:
but there is a dynamic shared memory feature inside driver and we do nothing in case when OP-TEE is built without this capability

>  I think it will not take long to take a look through all capabilities and make some checks.

if so, can we do it inside this pull request add checks for existed features?

Reply:
I expected those feature in this PR

Reply:
Added check for OPTEE_SMC_SEC_CAP_DYNAMIC_SHM. As far as I understand we don't rely on any other capabilities.